code/processes/eodprocess.q

```diff
@@ -0,0 +1,90 @@
+\d .eod
+
+taildir:hsym `$getenv`KDBTAIL;                                             /-load in taildir env variables
+hdbdir:hsym `$getenv`KDBHDB;                                               /-load in hdb env variables
+ignorelist:@[value;`ignorelist;`heartbeat`logmsg];                         /-list of tables to ignore
+taildbs:key taildir;                                                       /-list of tailDBs that need saved to HDB
+taildirs:();                                                               /-empty list to append tailDB paths to - to be used
+                                                                           / when HDB save is complete to delete tailDB partitions
+\d .
+
+savescompleted:0;                                                          /-variable to check how many tailDBs have been saved to HDB
+
+endofday:{[pt;procname]
+  /- function to trigger data load & save to HDB once endofday message is received from tailer(s)
+  .lg.o[`endofday;"end of day message received from ",string[procname]," - ",string[pt]];
+  loadandsave[pt;procname];
+  };
+
+loadandsave:{[pt;procname]
+  /- function to load TDB data and save data to HDB
+  taildir:`$1_string ` sv (.eod.taildir;procname;`$string pt);
+  .eod.taildirs,:taildir;
+  .lg.o[`load;"loading TDB data from ",string[procname]];
+  @[.Q.l; taildir; {[e] .lg.e[`load;"Failed to load TDB : ",e]}];
+
+  /-define list of tables to be saved to HDB
+  tablelist:tables[] except .eod.ignorelist;
+
+  /-save tables to HDB
+  savetables[.eod.hdbdir;pt;tablelist];
+  savescompleted+::1;
+
+  .lg.o[`eodcomplete;"end of day sort complete for ",string[procname]];
+
+  /-check if all eod saves have been completed, if so trigger savecomplete
+  if[savescompleted = count .eod.taildbs;savecomplete[pt;tablelist]];
+  };
+
+savecomplete:{[pt;tablelist]
+  /-function to add p attr to HDB tables, delete tailDBs
+  addpattr[.eod.hdbdir;pt;] each tablelist;
+  deletetaildb each .eod.taildirs;
+
+  /-reset savescompleted counter and .eod.taildirs
+  savescompleted::0;
+  .eod.taildirs:();
+  };
+
+upserttopartition:{[dir;pt;tabname]
+  /-create directory path data will be saved to
+  /-need to delete int col from tables before they can be saved to HDB
+  directory:(` sv .Q.par[dir;pt;tabname],`);
+  data:delete int from select from tabname;
```

Comment:
I think assigning data in this way increases the memory usage of the process significantly. This may be the best approach though as deleting the column on disk or using os deldir or something similar might be tricky as then you have to update .d and other issues. There may be a better solutions somewhere though.

Reply:
I believe the int column comes the subpartitions for each sym in the tailDB so that column gets added when the tailDB is loaded into memory. Because of this I couldn't find an easy way around it as it seems to come from the structure of the tailDB and just using `delete int from tabname` was giving a 'par error

Reply:
Oh forgot it is partitioned by int on disk so probably can't delete the int column. I will have a look at the savedown and see if there is I can get it to partition by sym without needing a duplicate int column.

Reply:
Haven't looked in depth here, but I believe it should be similar to how the classic TorQ sort process handles things if wdb is in partbyattr mode (where intraday data is partitioned).

The code involved in merging at EOD is in https://github.com/AquaQAnalytics/TorQ/blob/master/code/common/merge.q - might be worth having a look at that and seeing if we can reuse the code from there rather than re-implementing from scratch

Reply:
okay great I'll have a look at that, thanks Jonathon